### ISSUE TITLE
error: storage size of `ctx` isn't known

### ISSUE THREAD

**Issue by jakubklimek** [openlink/virtuoso-opensource#744]
Error when building current develop branch on Ubuntu 18.04:
```
/bin/bash ../../libtool  --tag=CC   --mode=compile gcc -DHAVE_CONFIG_H -I. -I../../libsrc/Dk    -fno-strict-aliasing -O2  -Wall  -DNDEBUG -DPOINTER_64   -I/home/klimekj/virtuoso-opensource/libsrc/Xml.new  -DOPENSSL_NO_KRB5 -Dlinux -D_GNU_SOURCE -DFILE64 -D_LARGEFILE64_SOURCE   -I../../libsrc -I../../libsrc/Dk -I../../libsrc/zlib -I. -I../../libsrc/langfunc -I../../libsrc/plugin -I../../libsrc/Tidy -I../../libsrc/Xml.new -I../../libsrc/odbcsdk/include -DVAD -DDBP -DBIF_XPER -DOPSYS=\"Linux\" -DHOST=\"x86_64-pc-linux-gnu\" -g -O2 -MT libwi_la-bif_crypto.lo -MD -MP -MF .deps/libwi_la-bif_crypto.Tpo -c -o libwi_la-bif_crypto.lo `test -f 'bif_crypto.c' || echo './'`bif_crypto.c
libtool: compile:  gcc -DHAVE_CONFIG_H -I. -I../../libsrc/Dk -fno-strict-aliasing -O2 -Wall -DNDEBUG -DPOINTER_64 -I/home/klimekj/virtuoso-opensource/libsrc/Xml.new -DOPENSSL_NO_KRB5 -Dlinux -D_GNU_SOURCE -DFILE64 -D_LARGEFILE64_SOURCE -I../../libsrc -I../../libsrc/Dk -I../../libsrc/zlib -I. -I../../libsrc/langfunc -I../../libsrc/plugin -I../../libsrc/Tidy -I../../libsrc/Xml.new -I../../libsrc/odbcsdk/include -DVAD -DDBP -DBIF_XPER -DOPSYS=\"Linux\" -DHOST=\"x86_64-pc-linux-gnu\" -g -O2 -MT libwi_la-bif_crypto.lo -MD -MP -MF .deps/libwi_la-bif_crypto.Tpo -c bif_crypto.c  -fPIC -DPIC -o .libs/libwi_la-bif_crypto.o
bif_crypto.c: In function ‘box_hmac’:
bif_crypto.c:184:12: error: storage size of ‘ctx’ isn’t known
   HMAC_CTX ctx;
            ^~~
bif_crypto.c:190:3: warning: ‘HMAC_Init’ is deprecated [-Wdeprecated-declarations]
   HMAC_Init (&ctx, key, box_length (key) - DV_STRINGP (key) ? 1 : 0, md);
   ^~~~~~~~~
In file included from /usr/include/openssl/hmac.h:13:0,
                 from /usr/include/openssl/ssl.h:56,
                 from http.h:39,
                 from bif_crypto.c:37:
/usr/include/openssl/hmac.h:28:1: note: declared here
 DEPRECATEDIN_1_1_0(__owur int HMAC_Init(HMAC_CTX *ctx, const void *key, int len,
 ^
bif_crypto.c:184:12: warning: unused variable ‘ctx’ [-Wunused-variable]
   HMAC_CTX ctx;
            ^~~
bif_crypto.c: In function ‘asn1_parse_to_xml’:
bif_crypto.c:351:13: warning: implicit declaration of function ‘d2i_ASN1_BOOLEAN’; did you mean ‘V_ASN1_BOOLEAN’? [-Wimplicit-function-declaration]
        ii = d2i_ASN1_BOOLEAN (NULL, (const unsigned char **)&opp, len + hl);
             ^~~~~~~~~~~~~~~~
             V_ASN1_BOOLEAN
bif_crypto.c:418:5: warning: implicit declaration of function ‘M_ASN1_OCTET_STRING_free’; did you mean ‘ASN1_OCTET_STRING_free’? [-Wimplicit-function-declaration]
     M_ASN1_OCTET_STRING_free (os);
     ^~~~~~~~~~~~~~~~~~~~~~~~
     ASN1_OCTET_STRING_free
bif_crypto.c:451:8: warning: implicit declaration of function ‘M_ASN1_INTEGER_free’; did you mean ‘ASN1_INTEGER_free’? [-Wimplicit-function-declaration]
        M_ASN1_INTEGER_free (bs);
        ^~~~~~~~~~~~~~~~~~~
        ASN1_INTEGER_free
bif_crypto.c:482:8: warning: implicit declaration of function ‘M_ASN1_ENUMERATED_free’; did you mean ‘ASN1_ENUMERATED_free’? [-Wimplicit-function-declaration]
        M_ASN1_ENUMERATED_free (bs);
        ^~~~~~~~~~~~~~~~~~~~~~
        ASN1_ENUMERATED_free
bif_crypto.c: In function ‘bio_to_strses’:
bif_crypto.c:726:33: error: dereferencing pointer to incomplete type ‘BIO {aka struct bio_st}’
   to_free = ((BUF_MEM *) out_bio->ptr)->data;
                                 ^~
bif_crypto.c: In function ‘bif_smime_sign’:
bif_crypto.c:927:21: error: dereferencing pointer to incomplete type ‘X509_STORE {aka struct x509_store_st}’
   if (store && store->objs)
                     ^~
bif_crypto.c:932:11: error: dereferencing pointer to incomplete type ‘X509_OBJECT {aka struct x509_object_st}’
    if (obj->type == X509_LU_X509)
           ^~
bif_crypto.c: In function ‘x509_certificate_verify_cb’:
bif_crypto.c:1184:18: error: dereferencing pointer to incomplete type ‘X509_STORE_CTX {aka struct x509_store_ctx_st}’
       switch (ctx->error)
                  ^~
bif_crypto.c: In function ‘BN_box’:
bif_crypto.c:1317:37: error: dereferencing pointer to incomplete type ‘BIGNUM {aka struct bignum_st}’
     buf = box_num ((unsigned long) x->d[0]);
                                     ^~
bif_crypto.c: In function ‘bif_get_certificate_info’:
bif_crypto.c:1501:41: error: dereferencing pointer to incomplete type ‘X509 {aka struct x509_st}’
  STACK_OF (X509_EXTENSION) * exts = cert->cert_info->extensions;
                                         ^~
bif_crypto.c:1513:5: warning: implicit declaration of function ‘M_ASN1_OCTET_STRING_print’; did you mean ‘ASN1_OCTET_STRING_set’? [-Wimplicit-function-declaration]
     M_ASN1_OCTET_STRING_print (mem, ex->value);
     ^~~~~~~~~~~~~~~~~~~~~~~~~
     ASN1_OCTET_STRING_set
bif_crypto.c:1513:39: error: dereferencing pointer to incomplete type ‘X509_EXTENSION {aka struct X509_extension_st}’
     M_ASN1_OCTET_STRING_print (mem, ex->value);
                                       ^~
bif_crypto.c:1540:11: error: dereferencing pointer to incomplete type ‘EVP_PKEY {aka struct evp_pkey_st}’
      if (k->type == EVP_PKEY_RSA)
           ^~
bif_crypto.c:1543:65: error: dereferencing pointer to incomplete type ‘RSA {aka struct rsa_st}’
   ret = list (3, box_dv_short_string ("RSAPublicKey"), BN_box (x->e), BN_box (x->n));
                                                                 ^~
bif_crypto.c:1551:65: error: dereferencing pointer to incomplete type ‘DSA {aka struct dsa_st}’
   ret = list (2, box_dv_short_string ("DSAPublicKey"), BN_box (x->pub_key));
                                                                 ^~
bif_crypto.c:1570:61: error: dereferencing pointer to incomplete type ‘X509_NAME {aka struct X509_name_st}’
  for (i = 0; NULL != subj && i < sk_X509_NAME_ENTRY_num(subj->entries); i++)
                                                             ^~
bif_crypto.c:1573:25: error: dereferencing pointer to incomplete type ‘X509_NAME_ENTRY {aka struct X509_name_entry_st}’
      n = OBJ_obj2nid (ne->object);
                         ^~
bif_crypto.c: In function ‘bif_base36dec’:
bif_crypto.c:1819:16: warning: initialization makes integer from pointer without a cast [-Wint-conversion]
   boxint str = bif_string_arg (qst, args, 0, "base36dec");
                ^~~~~~~~~~~~~~
bif_crypto.c:1821:17: warning: passing argument 1 of ‘strtoul’ makes pointer from integer without a cast [-Wint-conversion]
   ret = strtoul(str, NULL, 36);
                 ^~~
In file included from ../../libsrc/odbcsdk/include/iodbcunix.h:79:0,
                 from ../../libsrc/odbcsdk/include/sql.h:87,
                 from odbcinc.h:72,
                 from sqlnode.h:31,
                 from bif_crypto.c:28:
/usr/include/stdlib.h:180:26: note: expected ‘const char * restrict’ but argument is of type ‘boxint {aka long long int}’
 extern unsigned long int strtoul (const char *__restrict __nptr,
                          ^~~~~~~
Makefile:1306: recipe for target 'libwi_la-bif_crypto.lo' failed
make[3]: *** [libwi_la-bif_crypto.lo] Error 1
make[3]: Leaving directory '/home/klimekj/virtuoso-opensource/libsrc/Wi'
Makefile:2988: recipe for target 'install' failed
make[2]: *** [install] Error 2
make[2]: Leaving directory '/home/klimekj/virtuoso-opensource/libsrc/Wi'
Makefile:477: recipe for target 'install-recursive' failed
make[1]: *** [install-recursive] Error 1
make[1]: Leaving directory '/home/klimekj/virtuoso-opensource/libsrc'
Makefile:633: recipe for target 'install-recursive' failed
make: *** [install-recursive] Error 1
```

**Comment by jakubklimek** [comment]
Ahh, sorry, duplicate for https://github.com/openlink/virtuoso-opensource/issues/663

**Comment by AdrielFreud** [comment]
```
EVP_CIPHER_CTX *ctx;
ctx = EVP_CIPHER_CTX_new();
if(ctx == NULL){EVP_MD_CTX_free(ctx);}
```

**Comment by TallTed** [comment]
@AdrielFreud - Can you please provide some context for your comment? If it's intended to suggest a code change, a pull request would be helpful.

**Comment by AdrielFreud** [comment]
the error is caused by a memory failure or static allocation, you should try a dynamic allocation using some variants that the OPENSSL library offers.

**Comment by TallTed** [comment]
@AdrielFreud - I see. Perhaps you might make that suggestion on [the open issue](https://github.com/openlink/virtuoso-opensource/issues/663) #663, instead of [this closed duplicate](#issuecomment-390909596)?

**Comment by AdrielFreud** [comment]
I'm new to github, I do not know how to use it very well.
